core/src/main/scala/org/apache/spark/SparkThrowableHelper.scala

```diff
@@ -135,4 +140,59 @@ private[spark] object SparkThrowableHelper {
   def isInternalError(errorClass: String): Boolean = {
     errorClass == "INTERNAL_ERROR"
   }
+
+  def getMessage(e: SparkThrowable with Throwable, format: ErrorMessageFormat.Value): String = {
+    import ErrorMessageFormat._
+    format match {
+      case PRETTY => e.getMessage
+      case MINIMAL | STANDARD if e.getErrorClass == null =>
+        toJsonString { g =>
+          g.writeStartObject()
+          g.writeStringField("errorClass", "legacy")
+          g.writeObjectFieldStart("messageParameters")
+          g.writeStringField("message", e.getMessage)
+          g.writeEndObject()
+          g.writeArrayFieldStart("queryContext")
+          g.writeEndArray()
+          g.writeEndObject()
+        }
+      case MINIMAL | STANDARD =>
+        val errorClass = e.getErrorClass
+        assert(e.getParameterNames.size == e.getMessageParameters.size,
+          "Number of message parameter names and values must be the same")
+        toJsonString { generator =>
+          val g = if (format == STANDARD) generator.useDefaultPrettyPrinter() else generator
+          g.writeStartObject()
+          g.writeStringField("errorClass", errorClass)
+          val errorSubClass = e.getErrorSubClass
+          if (errorSubClass != null) g.writeStringField("errorSubClass", errorSubClass)
+          if (format == STANDARD) {
+            val errorInfo = errorClassToInfoMap.getOrElse(errorClass,
+              throw SparkException.internalError(s"Cannot find the error class '$errorClass'"))
+            g.writeStringField("message", errorInfo.messageFormat)
+          }
+          val sqlState = e.getSqlState
+          if (sqlState != null) g.writeStringField("sqlState", sqlState)
+          g.writeObjectFieldStart("messageParameters")
+          (e.getParameterNames zip e.getMessageParameters).foreach { case (name, value) =>
+            g.writeStringField(name, value)
+          }
+          g.writeEndObject()
+          g.writeArrayFieldStart("queryContext")
+          e.getQueryContext.map { c =>
+            g.writeStartObject()
+            g.writeStringField("objectType", c.objectType())
+            g.writeStringField("objectName", c.objectName())
+            val startIndex = c.startIndex() + 1
```

Comment:
does this mean it's inconsistent with the original text error message? cc @gengliangwang

Reply:
@cloud-fan Could you elaborate it a little bit more, please. What's the original error message? After https://github.com/apache/spark/pull/36651, the position is 1-based. And after my last commit, the indexes are 1-based. What do you mean by inconsistent?

Reply:
1 baed makes it consistent.